server/src/test/java/org/elasticsearch/search/query/SearchTimeoutTests.java

```diff
@@ -0,0 +1,289 @@
+                        // exposed on timeout triggered by the exitable directory reader or is that only the case when bulk scoring?
+                        final TermsEnum termsEnum = context.reader().terms("field").iterator();
+                        termsEnum.next();
+                        firstSegment = false;
+                        return new ConstantScoreScorer(this, score(), scoreMode, DocIdSetIterator.all(context.reader().maxDoc()));
+                    }
+
+                    @Override
+                    public boolean isCacheable(LeafReaderContext ctx) {
+                        return false;
+                    }
+                };
+            }
+
+            @Override
+            public String toString(String field) {
+                return "match_all";
+            }
+
+            @Override
+            public boolean equals(Object o) {
+                return sameClassAs(o);
+            }
+
+            @Override
+            public int hashCode() {
+                return classHash();
+            }
+
+            @Override
+            public void visit(QueryVisitor visitor) {
+                visitor.visitLeaf(this);
+            }
+        };
+        {
+            TestSearchContext context = new TestSearchContext(null, indexShard, newContextSearcher(reader));
+            context.setTask(new SearchShardTask(123L, "", "", "", null, Collections.emptyMap()));
+            context.parsedQuery(new ParsedQuery(query));
+            context.setSize(size);
+            QueryPhase.executeQuery(context);
+            assertFalse(context.queryResult().searchTimedOut());
+            assertEquals(numDocs, context.queryResult().topDocs().topDocs.totalHits.value);
+            assertEquals(size, context.queryResult().topDocs().topDocs.scoreDocs.length);
+        }
+        {
+            pulledFirstScorer[0] = false;
+            TestSearchContext context = new TestSearchContext(null, indexShard, newContextSearcher(reader)) {
+                @Override
+                public long getRelativeTimeInMillis() {
+                    return pulledFirstScorer[0] ? 1L : 0L;
+                }
+            };
+            context.setTask(new SearchShardTask(123L, "", "", "", null, Collections.emptyMap()));
+            context.parsedQuery(new ParsedQuery(query));
+            context.setSize(size);
+            QueryPhase.executeQuery(context);
+            assertTrue(context.queryResult().searchTimedOut());
+            int firstSegmentMaxDoc = reader.leaves().get(0).reader().maxDoc();
+            assertEquals(Math.min(2048, firstSegmentMaxDoc), context.queryResult().topDocs().topDocs.totalHits.value);
+            assertEquals(Math.min(size, firstSegmentMaxDoc), context.queryResult().topDocs().topDocs.scoreDocs.length);
+        }
+    }
+
+    public void testBulkScorerTimeout() throws IOException {
```

Comment:
There are many other places, I just don't know that we can reliably test them all. It is fascinating that we have so little test coverage so far around timeout support :)

Reply:
There is more support for search cancellations. Timeout is a specific query Cancellation that needs to be handled different by the index searcher, like we need to return the collected results until the timeout in order to return partial results. I think these tests achieve that goal for hits and aggregations.

Reply:
I have expanded them to cover when a Scorer throws timeout (via ExitableDirectoryReader) when pulling terms enum or points.

I plan on expanding this to vectors but it's more realistic to write dfs phase tests, as that is where such timeout may happen, if it only wasn't that the dfs phase does not support timeouts at all currently, which is something that I am planning on fixing and writing tests for.

Reply:
Also, I looked at timeouts possibly thrown by createWeight, and I could not find a scenario where that does happen in practice. I am reluctant to simulate that scenario if it never happens in reality.

When it comes to rewrite, it happens outside of the search call, hence we don't catch the timeout exception properly, but we also won't get it, because we set the timeout runnable after rewrite has been done. This needs fixing for both query phase and dfs phase. I believe it's more important for the knn query (which runs in dfs) as its rewrite is the heavy work that knn does.